File: flask-server/blueprints/categories.py
from flask import Blueprint, jsonify, request, session
import requests, logging

categories_bp = Blueprint('categories', __name__)

# Configure logging
logging.basicConfig(level=logging.INFO)

@categories_bp.route('/genres')
def GetGenres():
    genres = []
    response = requests.get(f'https://wolnelektury.pl/api/genres/')
    if response.status_code == 200:
        genresResponse = response.json()
        genres.extend(genresResponse)
        logging.info(f"Fetched {len(genres)} genres")
    else:
        logging.error(f"Failed to fetch genres")
    return jsonify(genres)

@categories_bp.route('/sendGenre', methods=['POST'])
def sendGenre():
    data = request.get_json()
    genreName = data.get('genre')
    if not genreName:
        return jsonify({'error': 'Brak gatunku'}), 400
    session['current_genre'] = genreName
    return jsonify({'message': f'Zapisano gatunek: {genreName}'}), 200

@categories_bp.route('/genreBooks')
def genreBooks():
    genreName = session.get('current_genre')
    books = []
    response = requests.get(f'https://wolnelektury.pl/api/genres/{genreName}/books/')
    if response.status_code == 200:
        genre_books = response.json()
        books.extend(genre_books)
        logging.info(f"Fetched {len(genre_books)} books for genre {genreName}")
    else:
        logging.error(f"Failed to fetch books for genre {genreName}")
    return jsonify(books)

File: client/src/components/Categories.js
import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import '../styles/Categories.css';

function GetGenres() {
  const [genres, setGenres] = useState([]);
  const [selectedGenre, setSelectedGenre] = useState('artykul');
  const [books, setBooks] = useState([]);
  const [error, setError] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const navigate = useNavigate();

  // Pobieranie gatunków
  useEffect(() => {
      fetch('/genres')
          .then(response => response.json())
          .then((data) => {
            setGenres(data); 
            const defaultGenre = data.find((genre) => genre.name === 'Artykuł');
            if (defaultGenre) {
              setSelectedGenre(defaultGenre.slug); 
            }
          })
          .catch(error => handleError(error));
  }, []);

  // Pobieranie książek przy zmianie gatunku
  useEffect(() => {
      if (selectedGenre) {
        setIsLoading(true); 
        setBooks([]); 

          fetch('/sendGenre', {
              method: 'POST',
              headers: { 'Content-Type': 'application/json' },
              body: JSON.stringify({ genre: selectedGenre })
          })
          .then(() => fetch('/genreBooks'))
          .then(response => response.json())
          .then((data) => {
            //setBooks(data.slice(0, 18));
            setBooks(data);
            setIsLoading(false); // Koniec ładowania
          })
          .catch((error) => {
            handleError(error);
            setIsLoading(false); // Koniec ładowania
          });
          
      }
  }, [selectedGenre]);

  const handleError = (error) => {
      console.error('Error:', error);
      alert('Wystąpił błąd podczas pobierania danych.');
      setError(true);
  };

  const handleGenreClick = (genreSlug) => {
      setSelectedGenre(genreSlug);
  };

  if (error) {
      return <div className="errorCategories"><h1>Strona Tymczasowo Niedostępna</h1></div>;
  }

  return (
      <div className="main">
          <div className="collectionCategories">             
              <div className="genres">
                <div className='genres-title'>
                  <h1>Dostępne gatunki</h1>
                </div>               
                <div className='genres-module'>
                  {genres.map((genre, index) => (
                      <div key={index} className="genre">
                          <button onClick={() => handleGenreClick(genre.slug)}>
                              {genre.name}
                          </button>
                      </div>
                  ))}
                </div>
              </div>
              {selectedGenre && (
                    <div className="books-section">
                      <h3>Książki z gatunku: {
                        genres.find(g => g.slug === selectedGenre)?.name
                      }</h3>
                      {isLoading ? (
                      <div className="loading-message">
                        Ładowanie książek...
                      </div>
                      ) : books.length > 0 ? (
                        <div className="books">
                          {books.map((book, index) => (
                            <div key={index} className="book">
                              <div onClick={() => navigate(`/book/${book.slug}`)} style={{ cursor: 'pointer' }}>
                                <div className="book-cover-container">
                                  <img 
                                    src={`https://wolnelektury.pl/media/${book.cover}`} 
                                    alt={book.title} 
                                    className="book-cover" 
                                  />
                                  <div className="book-title">{book.title}</div>
                                </div>
                              </div>
                            </div>
                          ))}
                        </div>
                      ) : (
                        <p>Brak dostępnych książek</p>
                      )}
                    </div>
                  )}
          </div>
      </div>
  )

  
}

export default GetGenres;